### PR TITLE
all: implement EIP-7002 and EIP-7251, execution layer triggered withdrawal and consolidation requests

### PR DESCRIPTION
**Pull request by lightclient**
~~*This PR builds upon the 6110 impl #29431*~~ merged!

This PR implements [EIP-7002](https://eips.ethereum.org/EIPS/eip-7002): Execution layer triggerable exits and [EIP-7251](https://eips.ethereum.org/EIPS/eip-7251): Increase the MAX_EFFECTIVE_BALANCE

### CODE REVIEW

File: consensus/consensus.go [hunk 1]
```diff
@@ -97,7 +97,7 @@ type Engine interface {
 	// Note: The block header and state database might be updated to reflect any
 	// consensus rules that happen at finalization (e.g. block rewards).
 	FinalizeAndAssemble(chain ChainHeaderReader, header *types.Header, state *state.StateDB, txs []*types.Transaction,
-		uncles []*types.Header, receipts []*types.Receipt, withdrawals []*types.Withdrawal) (*types.Block, error)
+		uncles []*types.Header, receipts []*types.Receipt, withdrawals []*types.Withdrawal, exits []*types.Exit) (*types.Block, error)
```
Comment (MariusVanDerWijden) [inline consensus/consensus.go:100]: we should think about turning this into some struct at some point.
Might actually be nicer for downstream projects that implement their own consensus engines as well

**Comment by MariusVanDerWijden** [comment]
Is there an upper limit on exits?

File: consensus/consensus.go (hunk 1 above)
Comment (lightclient) [reply consensus/consensus.go:100]: Yeah, we'll also have deposits here for 6110, so I think it's time to figure out a better format for this.

**Comment by lightclient** [comment]
The max number of deposits is in the system contract and is 16.

File: consensus/consensus.go (hunk 1 above)
Comment (itsdevbear) [reply consensus/consensus.go:100]: maybe time to make a formal BlockBody like in Prysm?

Comment (fjl) [reply consensus/consensus.go:100]: It was changed on master branch to use `types.Body`.
All existing tracers in eth/tracers/native have been rewritten to use the new hook system.

**Review by Imrall — APPROVED** [review]

**Review by holiman — APPROVED** [review]
Look about right to me.

File: core/state_processor.go [hunk 2]
```diff
@@ -275,3 +280,81 @@ func ParseDepositLogs(logs []*types.Log, config *params.ChainConfig) (types.Requ
...
+		To:        &params.WithdrawalRequestsAddress,
+	}
+	vmenv.Reset(NewEVMTxContext(msg), statedb)
+	statedb.AddAddressToAccessList(params.WithdrawalRequestsAddress)
+	ret, _, _ := vmenv.Call(vm.AccountRef(msg.From), *msg.To, msg.Data, 30_000_000, common.U2560)
+	statedb.Finalise(true)
+
+	// Parse out the exits.
+	var reqs types.Requests
+	for i := 0; i < len(ret)/76; i++ {
```
Comment (MariusVanDerWijden) [inline core/state_processor.go:307]: Should we check that ret is % 76 == 0 here or just accept anything thats returned from the contract?

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM

**Comment by fjl** [comment]
Note that work on the spec is still ongoing, and we will have some simplifications in soon that will invalidate most of this PR.

**Comment by fjl** [comment]
I have created a new implementation in #30571 since it is so much simpler now. There was just no point trying to rebase this one.

**Pull request closed without merging**